skills_ml/algorithms/embedding/train.py

```diff
@@ -68,20 +69,13 @@ def __init__(
             logging.warning("Current gensim doc2vec doesn't support online batch learning. Use generic learning instead.")
             logging.warning("Training on too large corpus might cause serious memory leaks!")
 
-        if source not in ['nlx', 'cb', 'all']:
-            raise ValueError('"{}" is an invalid source!'.format(source))
-
-        if not isinstance(quarters, list):
-            raise TypeError('quarters should be a list of string, e.g. ["2011Q1", "2011Q2"]')
-
+        self.corpus_generator = corpus_generator
         self.s3_conn = s3_conn
-        self.quarters = quarters
-        self.jp_s3_path = jp_s3_path
+        self.quarters = corpus_generator.quarters
```

Comment:
You may not even need to do this. It looks like the only use of `self.quarters` is the fact that we save it in the metadata. You could just access self.corpus_generator.quarters when it is time to do so.

However, I have a different idea. Instead of making the model trainer too aware of specific pieces of metadata that its data source has, what if the corpus generator just had a .metadata property that the model trainer saves along with its metadata? That we can keep it in a different place (eg. `metadata['corpus']['quarters']` instead of `metadata['quarters']`) from metadata that is more geared towards the model itself. And, we can add metadata to the corpus at will without needing to update it here.

Reply:
make sense. so if I understand right, both corpus generator and trainer will have `metadata` property, but one in corpus generator has metadata about the data source and the other one is about the model parameters.

Reply:
Basically. If there's one thing that all of the nodes in these DAGs should have in common, it's that they should each have some kind of metadata, because we want to attach this metadata wherever the outputs are saved.

JobPostingGenerator -> Quarters, Data Source
CorpusCreator -> The transformation function (even just the name of it)
Trainer -> A bunch of model parameters

Each step should have the metadata of all the steps for it, so the JobPostingGenerator may have a list like this:

```
{
  'job_posting_generator': { 'quarters': ['2014Q1', '2014Q2'], 'data_sources': 'NLX' }
]
```

The metadata gets passed from the JobPostingGenerator to whatever comes after it, with its own metadata dictionary merged in. So then the CorpusCreator would have this:

```
  'job_posting_generator': { 'quarters': ['2014Q1', '2014Q2'], 'data_sources': 'NLX' }
  'corpus_creator': 'Word2VecGensimCorpusCreator'
``` 


In the same fashion, the Trainer end up with this:

```
  'job_posting_generator': { 'quarters': ['2014Q1', '2014Q2'], 'data_sources': 'NLX' }
  'corpus_creator': 'Word2VecGensimCorpusCreator'
  'embedding_trainer': { 'hyperparameter1': 'value1', 'hyperparameter2': 'value2', ...}
``` 

Ideally, this would be done in a mostly automatic way. No step should need to reach in and grab specific pieces of metadata from a prior step, but rather just bundle that metadata in its entirety whenever it persists its own.